Content.Shared/Damage/DamageableSystem.cs

```diff
@@ -0,0 +1,313 @@
+using Robust.Shared.Log;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Damage
+{
+    public class DamageableSystem : EntitySystem
+    {
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+        public override void Initialize()
+        {
+            SubscribeLocalEvent<DamageableComponent, ComponentInit>(DamageableInit);
+            SubscribeLocalEvent<DamageableComponent, TryChangeDamageEvent>(TryChangeDamage);
+            SubscribeLocalEvent<DamageableComponent, ComponentHandleState>(DamageableHandleState);
+            SubscribeLocalEvent<DamageableComponent, ComponentGetState>(DamageableGetState);
+        }
+
+        /// <summary>
+        ///     Initialize a damageable component
+        /// </summary>
+        private void DamageableInit(EntityUid uid, DamageableComponent component, ComponentInit _)
+        {
+            // Get resistance set, if any was specified.
+            if (component.ResistanceSetID != null)
+            {
+                _prototypeManager.TryIndex(component.ResistanceSetID, out component.ResistanceSet);
+            }
+
+            // Note that component.DamageContainerID may be null despite being a required data field. In particular,
+            // this can happen when adding this component via ViewVariables.
+
+            if (component.DamageContainerID == null ||
+                !_prototypeManager.TryIndex<DamageContainerPrototype>(component.DamageContainerID, out var damageContainerPrototype) ||
+                damageContainerPrototype.SupportAll )
+            {
+                // Either the container should support all damage types, or no valid DamageContainerID was given (for
+                // which we default to support all). Add every single damage type to our component:
+                foreach (var type in _prototypeManager.EnumeratePrototypes<DamageTypePrototype>())
+                {
+                    component.DamagePerType.Add(type.ID, 0);
+                }
+
+                DamageChanged(uid, component, false);
+                return;
+            }
+
+            // A valid damage container ID was provided. Initialize damage dictionary, using the types from the damage
+            // container prototype
+            component.DamagePerType = new(damageContainerPrototype.SupportedTypes.Count);
+            foreach (var type in damageContainerPrototype.SupportedTypes)
+            {
+                component.DamagePerType.Add(type, 0);
+            }
+
+            // Then also add the supported damage groups
+            foreach (var groupID in damageContainerPrototype.SupportedGroups)
+            {
+                var group = _prototypeManager.Index<DamageGroupPrototype>(groupID);
+                foreach (var type in group.DamageTypes)
+                {
+                    component.DamagePerType.TryAdd(type, 0);
+                }
+            }
+
+            DamageChanged(uid, component, false);
```

Comment:
Currently initialising with damage isn't supported, so yeah it should just be removed (which I have now done). However making this change had a sort of domino effect that required several other changes that people may want to look at. So here is the reasoning behind those:

This dirty-on-init behaviour was around before even the damageable refactor and as far as I can tell the MobStateComponent appears to have inadvertently become reliant on the damage-changed event/message to effectively do some initialisation for it? So I've now also made changes to `MobStateComponent` to fix that.

AFAICT these changes accidentally fixed an error in `DisposalUnitTest` where the test was encountering an error when spawning entities (specifically the human dummy with a `MobStateComponent`). Because this was done in an un-awaited async function the errors were never raised and the test appeared to be succeeding? My changes in `MobStateComponent` fixed that error, so the test was actually running and getting to the actuall test assertions, which were failing. So I've also made changes to the disposal unit test so that it actually succeeds now.